warehouse/query-core/src/main/java/datawave/query/function/LimitFields.java

```diff
@@ -87,145 +58,186 @@ public Multimap<String,Attribute<? extends Comparable<?>>> load(String key) {
-            Multimap<String,Attribute<? extends Comparable<?>>> hitMap = hits.getUnchecked(keyNoGrouping);
-            for (String keyWithGrouping : hitMap.keySet()) {
-                for (Attribute<? extends Comparable<?>> value : hitMap.get(keyWithGrouping)) {
-                    // if(limit <= 0) break; // comment this line if you want to get ALL hits even if the limit is exceeded
-                    reducedMap.put(keyWithGrouping, value);
-                    limit--;
+        
+        // second pass is to set any misses back to be kept if the limit allows
+        for (Map.Entry<String,Attribute<? extends Comparable<?>>> de : document.entrySet()) {
+            String keyWithGrouping = de.getKey();
+            String keyNoGrouping = removeGrouping(keyWithGrouping);
+            
+            // look for the key without the grouping context
+            if (this.limitFieldsMap.containsKey(keyNoGrouping)) {
+                int limit = this.limitFieldsMap.get(keyNoGrouping);
+                
+                // short circuit if we are not actually limiting this field.
+                // this is keeping with the original logic where a negative limit means to keep only hits
+                if (limit <= 0) {
+                    continue;
                 }
-            }
-            Multimap<String,Attribute<? extends Comparable<?>>> missMap = misses.getUnchecked(keyNoGrouping);
-            for (String keyWithGrouping : missMap.keySet()) {
-                for (Attribute<? extends Comparable<?>> value : missMap.get(keyWithGrouping)) {
-                    if (limit <= 0)
-                        break;
-                    reducedMap.put(keyWithGrouping, value);
-                    limit--;
+                
+                int keepers = countKeepersForFieldMap.get(keyNoGrouping);
+                int missesRemaining = countMissesRemainingForFieldMap.get(keyNoGrouping);
+                int missesToSet = Math.min(limit - keepers, missesRemaining);
+                
+                // if we have misses yet to keep
+                if (missesToSet > 0) {
+                    boolean foundMiss = false;
+                    Attribute<?> attr = de.getValue();
+                    if (attr instanceof Attributes) {
+                        Attributes attrs = (Attributes) attr;
+                        Set<Attribute<? extends Comparable<?>>> attrSet = attrs.getAttributes();
+                        
+                        for (Attribute<? extends Comparable<?>> value : attrSet) {
+                            // if this was an attribute previously set to not keep, then it is one of the misses (not a hit)
+                            if (!value.isToKeep()) {
+                                value.setToKeep(true);
+                                keepers++;
+                                missesRemaining--;
+                                missesToSet--;
+                                attributesToDrop--;
```

Comment:
Do we need to set `foundMiss = true;` here?

Reply:
oh crud... I did have that in there.  good catch!